readthedocs/search/api/v3/backend.py

```diff
@@ -0,0 +1,225 @@
+from functools import cached_property
+from itertools import islice
+
+from readthedocs.builds.models import Version
+from readthedocs.projects.models import Project
+from readthedocs.search.api.v3.queryparser import SearchQueryParser
+from readthedocs.search.faceted_search import PageSearch
+
+
+class Backend:
```

Comment:
I have no idea how to name this "thing", backend was the thing the came to my head, my I don't like it. This is just an abstraction that we can re-use in the api and the global search view.

Reply:
`SearchExecuter` or `SearchWrapper` perhaps? I don't fully understand what it's doing, but we can definitely come up with a better name, and it should have `Search` in it as well.